### PR TITLE
[system-probe] Do not use `tcp_cleanup_rbuf` for tcp receive tracking

### PR DESCRIPTION
**Pull request by hmahmood**
### What does this PR do?

Use kprobes on `tcp_recvmsg` and `tcp_read_sock` instead of `tcp_cleanup_rbuf` for tcp receive tracking. `tcp_cleanup_rbuf` is also called for retransmits, which would overcount tcp receives.

### Reviewer's Checklist

- [x] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [x] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [x] Changed code has automated tests for its functionality.
- [x] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [x] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

File: pkg/network/ebpf/c/tcp-recv.h [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+
+    bpf_map_delete_elem(&tcp_recvmsg_args, &pid_tgid);
+
+    struct sock *skp = *skpp;
+    if (!skp) {
+        return 0;
+    }
+
+    int recv = PT_REGS_RC(ctx);
+    if (recv < 0) {
```
Comment (brycekahle) [inline pkg/network/ebpf/c/tcp-recv.h:47]: Can we do this first? It could save a lot of work.

File: pkg/network/ebpf/c/tcp-recv.h [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+
+SEC("kretprobe/tcp_read_sock")
+int kretprobe__tcp_read_sock(struct pt_regs *ctx) {
+    u64 pid_tgid = bpf_get_current_pid_tgid();
+    struct sock **skpp = (struct sock**) bpf_map_lookup_elem(&tcp_recvmsg_args, &pid_tgid);
+    if (!skpp) {
+        return 0;
+    }
+
+    bpf_map_delete_elem(&tcp_recvmsg_args, &pid_tgid);
```
Comment (nplanel) [inline pkg/network/ebpf/c/tcp-recv.h:70]: I would copy the pointer on the stack (line 72) before deleting the map entry

File: pkg/network/ebpf/c/tcp-recv.h [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+
+SEC("kretprobe/tcp_recvmsg")
+int kretprobe__tcp_recvmsg(struct pt_regs *ctx) {
+    u64 pid_tgid = bpf_get_current_pid_tgid();
+    struct sock **skpp = (struct sock**) bpf_map_lookup_elem(&tcp_recvmsg_args, &pid_tgid);
+    if (!skpp) {
+        return 0;
+    }
+
+    bpf_map_delete_elem(&tcp_recvmsg_args, &pid_tgid);
```
Comment (nplanel) [inline pkg/network/ebpf/c/tcp-recv.h:39]: I would copy the pointer on the stack (line 41) before deleting the map entry

**Review by nplanel — CHANGES_REQUESTED** [review]
We probably need a tcp splice() unit test to be safe in our CI, but we can address that in another PR

File: pkg/network/ebpf/c/tcp-recv.h (hunk 2 above)
Comment (nplanel) [reply pkg/network/ebpf/c/tcp-recv.h]: Even if it's working, (and would probably possibly failed on high load/high nb cpus TBC), the semantic is wrong, would you use a pointer after free() ? Would you guarantee that the kernel will not pollute the value at the end of delete call ? Still surprise the verifier don't check that. PR with the remaining fixes https://github.com/DataDog/datadog-agent/pull/12687

**Review by nplanel — APPROVED** [review]

**Pull request merged**